torch/jit/_recursive.py

```diff
@@ -140,13 +137,23 @@ def infer_type(name, item):
 
     for name in constants_set:
         if name in added_names:
-            # XXX: It is possible for something to be in the constants set but
-            # also in the parameters/buffers. This happens in BatchNorm as a
-            # hack to support optional parameters.
+            # TODO: We should really error in this case, but its bc-breaking so
+            # we need to warn for at least one release
+            if name in nn_module._modules:
+                hint = "submodule"
+            elif name in nn_module._buffers:
+                hint = "buffer"
+            elif name in nn_module._parameters:
+                hint = "parameter"
+            else:
+                raise AssertionError("added_names must be submodule, parameter, or buffer")
```

Comment:
Attributes are handled separately (and last), because unlike modules/buffers/parameters they don't have an actual definition in the nn.Module API, they are just Python attributes. The purpose of added_names is to make sure we don't re-add attributes that already exist as modules/buffers/parameters, so by definition it can't be an attr.